### PR TITLE
EVG-2235 improve performance of bulk task restart

### PR DESCRIPTION
**Pull request by john-m-liu**

### CODE REVIEW

File: units/tasks_restart.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+type restartTasksJob struct {
+	job.Base  `bson:"job_base" json:"job_base" yaml:"job_base"`
+	logger    grip.Journaler
+	startTime time.Time
+	endTime   time.Time
+	user      string
+	opts      model.RestartTaskOptions
+}
+
+// NewTasksRestartJob creates a job to restart failed tasks in a time range
```
Comment (tychoish) [inline units/tasks_restart.go]: - register the job. 
- populate the job.Base: JobType field, etc.) 
- make the job serializable (export fields, add tags).

File: units/tasks_restart.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		opts:      opts,
+	}
+	job.SetID(util.RandomString())
+	return &job
+}
+
+func (j *restartTasksJob) Run() {
+	results, err := model.RestartFailedTasks(j.startTime, j.endTime, j.user, j.opts)
+	if err != nil {
+		j.logger.Error(errors.Wrap(err, "error restarting failed tasks"))
```
Comment (tychoish) [inline units/tasks_restart.go]: job.Base has an .AddErrors() method which is nice.

File: rest/data/interface.go [hunk 3]
```diff
@@ -139,7 +140,7 @@ type Connector interface {
 	SetBannerTheme(string, *user.DBUser) error
 	// SetAdminBanner sets set the service flags in the system-wide settings document
 	SetServiceFlags(admin.ServiceFlags, *user.DBUser) error
-	RestartFailedTasks(time.Time, time.Time, string, model.RestartTaskOptions) (*restModel.RestartTasksResponse, error)
+	RestartFailedTasks(time.Time, time.Time, string, model.RestartTaskOptions, evergreen.Environment) (*restModel.RestartTasksResponse, error)
```
Comment (tychoish) [inline rest/data/interface.go]: I'd make it the first argument, rather than the last.

File: units/tasks_restart.go (hunk 1 above)
Comment (john-m-liu) [reply units/tasks_restart.go]: done

File: units/tasks_restart.go (hunk 2 above)
Comment (john-m-liu) [reply units/tasks_restart.go]: switched to this

File: units/tasks_restart.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/mongodb/grip/logging"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+const restartTasksJobName = "restart-tasks"
+
+func init() {
+	registry.AddJobType(restartTasksJobName,
+		func() amboy.Job { return &restartTasksJob{} })
```
Comment (tychoish) [inline units/tasks_restart.go]: the job type isn't properly set by default.

File: units/tasks_restart.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		EndTime:   endTime,
+		User:      user,
+		Opts:      opts,
+	}
+	job.JobType = amboy.JobType{
+		Name:    restartTasksJobName,
+		Version: 0,
+		Format:  amboy.BSON,
+	}
+	job.SetID(util.RandomString())
```
Comment (tychoish) [inline units/tasks_restart.go]: I would have the ID be some combination of the inputs to this function.

File: units/tasks_restart.go (hunk 4 above)
Comment (john-m-liu) [reply units/tasks_restart.go]: changed to call the constructor with dummy values

File: units/tasks_restart.go (hunk 5 above)
Comment (john-m-liu) [reply units/tasks_restart.go]: added

**Pull request merged**